enumerative.ml

```diff
@@ -37,24 +37,48 @@ end)
 
 module SI = Set.Make (struct
     type t = int
-    let compare = Pervasives.compare
+    let compare = Stdlib.compare
   end)
 
 module SLI = Set.Make (struct
     type t = int list
-    let compare = Pervasives.compare
+    let compare = Stdlib.compare
   end)
 
 module TMap = Map.Make (Term)
 
-type state = int array
+module State : sig
+  type t
+  val make : int -> int -> t
+  val copy : t -> t
+  val length : t -> int
+  val get : t -> int -> int
+  val set : t -> int -> int -> unit
+  val fold_left : ('a -> int -> 'a) -> 'a -> t -> 'a
+  val iteri : (int -> int -> unit) -> t -> unit
+  val as_array : t -> int array
+end = struct
+  type t = int array
+  let copy a = Obj.(obj (with_tag abstract_tag (repr a)))
+  let make len elt = copy (Array.make len elt)
+  let length = Array.length
+  let get = Array.get
+  let set = Array.set
+  let fold_left = Array.fold_left
+  let iteri = Array.iteri
+  let as_array x = x
+end
```

Comment:
What is the reasoning behind this? Using `Obj.magic` below seems highly dangerous